agent/connect/ca/provider_consul.go

```diff
@@ -290,6 +410,96 @@ func (c *ConsulProvider) Sign(csr *x509.CertificateRequest) (string, error) {
 	return buf.String(), nil
 }
 
+// SignIntermediate will validate the CSR to ensure the trust domain in the
+// URI SAN matches the local one and that basic constraints for a CA certificate
+// are met. It should return a signed CA certificate with a path length constraint
+// of 0 to ensure that the certificate cannot be used to generate further CA certs.
+func (c *ConsulProvider) SignIntermediate(csr *x509.CertificateRequest) (string, error) {
+	idx, providerState, err := c.getState()
+	if err != nil {
+		return "", err
+	}
+
+	if len(csr.URIs) < 1 {
+		return "", fmt.Errorf("intermediate CSR has no URIs")
+	}
+	certURI, err := connect.ParseCertURI(csr.URIs[0])
+	if err != nil {
+		return "", err
+	}
+
+	// Verify that the trust domain is valid
+	if !c.spiffeID.CanSign(certURI) {
```

Comment:
This feels like it can be moved up into the sign intermediate RPC endpoint once we have that (so this logic doesn't have to be duplicated in each provider), but I just added it here for now. The same goes for the validation of the incoming cert in SetIntermediate.

Reply:
As noted above we should probably also validate that the CSR is for a CA cert. I realise that in our current model the URI is only used as a CA cert but it feels better to be explicit and conservative rather than leave things open to potential abuses we've not considered.